### PR TITLE
pid1: add env var to override default mount rate limit burst

### PR DESCRIPTION
**Pull request by bluca**
I am hitting the rate limit on a busy system with low resources, and
it stalls the boot process which is Very Bad (TM).
Add an option to the manager to allow overriding locally.

### CODE REVIEW

**Comment by poettering** [comment]
Hmm, if the ratelimit is too low for a real-life system that is not in a bad state, then let's just bump the limit for good?

i.e. thse ratelimit should be high enough to not cause issues in real-life situations.

**Comment by bluca** [comment]
> Hmm, if the ratelimit is too low for a real-life system that is not in a bad state, then let's just bump the limit for good?
> 
> i.e. thse ratelimit should be high enough to not cause issues in real-life situations.

I don't know, this was added in v248 and IIRC nobody ever complained, so I didn't want to change the default and break booting for other cases. In the end it's really difficult to say, as it's very dependent from the local environment and what you are willing to risk - rate limiting reduces the chance of race conditions, so if you don't care about counting nanoseconds at boot you'll want it, but in my case I can't really afford it. In short, it seems to me something that would be good to have configurable.

**Comment by bluca** [comment]
changed to env var as requested

File: src/core/main.c [hunk 1]
```diff
@@ -2449,6 +2457,14 @@ static void reset_arguments(void) {
         arg_random_seed_size = 0;
         arg_clock_usec = 0;
 
+        arg_mount_rate_limit_burst = DEFAULT_MOUNT_RATE_LIMIT_BURST;
+        const char *e = secure_getenv("SYSTEMD_DEFAULT_MOUNT_RATE_LIMIT_BURST");
```
Comment (poettering) [inline src/core/main.c]: reset_arguments() is supposed to reset everything to defaults, i.e. the  secure_getenv() check and parsing should be moved out of here and instead into parse_config_file()

File: src/core/mount.c [hunk 2]
```diff
@@ -1948,7 +1948,7 @@ static void mount_enumerate(Manager *m) {
                         goto fail;
                 }
 
-                r = sd_event_source_set_ratelimit(m->mount_event_source, 1 * USEC_PER_SEC, 5);
+                r = sd_event_source_set_ratelimit(m->mount_event_source, 1 * USEC_PER_SEC, m->default_mount_rate_limit_burst ?: 5);
```
Comment (poettering) [inline src/core/mount.c]: Please drop the ternary op. First of all, you don't need it, the field should be set to 5 already by default anyway. Secondly, you take away the ability to turn off the ratelimiting by this. And I think it makes a ton of sense to be able to allow users to turn it off if they want. Hence, by just dropping the `?: 5` you lose nothing but gain simplicity and usefulness.

File: src/core/main.c [hunk 3]
```diff
@@ -2449,6 +2457,14 @@ static void reset_arguments(void) {
...
+        if (e) {
+                r = safe_atou(e, &arg_mount_rate_limit_burst);
+                if (r < 0)
+                        log_debug("Invalid value in $SYSTEMD_DEFAULT_MOUNT_RATE_LIMIT_BURST, ignoring: %s", e);
+        }
```
Comment (poettering) [inline src/core/main.c]: quite frankly, I'd drop this all here. Just read the env var directly in mount_enumerate(), that's the only place you need the variable, and this is not performance sensitive, given that the function is only called during initialization once. It becomes vastly simpler then and a very local to where the value is consumed.

Comment (bluca) [reply src/core/main.c]: ok, done

**Comment by poettering** [comment]
looks great, thanks! so much simpler!

**Comment by poettering** [comment]
(btw, I think it would be fantastic to bump the defaults too, if we can find some limits that are as low as you can get away with)

**Comment by keszybz** [comment]
Hmm, is this really suitable for a backport?

**Comment by bluca** [comment]
The test change commit definitely.

The env var I'd say would be good too, as it's a workaround for an issue introduced in 249, so I definitely plan to use it on top of 252, so would be nice to have it upstream

**Pull request merged**